### ISSUE TITLE
Pre-trained classifier

### ISSUE THREAD

**Issue by ayush1801** [rosinality/semantic-pyramid-pytorch#3]
In the pre-trained classifier, why the output layer is not the softmax classifier for classifying 365 labels? Also, are the classifier layers also pre-trained in VGG classifier?

**Comment by ayush1801** [comment]
@rosinality I am curious about why you haven't changed your final classifier layer of the pre-trained network (vgg16) to output 365 classes instead of 1000 classes. Is there any specific reason to this? It would be great if you could enlighten about this aspect.

**Comment by rosinality** [comment]
If you want to use label of places 365, then you will need to train or finetune vgg networks. Instead this model/implementation uses pretrained model directly. (Convenient)

**Comment by ayush1801** [comment]
Thanks @rosinality. Then, Are you feeding the class labels directly into the Generator and Discriminator? I thought that you might be feeding the labels predicted by the classifier.

**Comment by rosinality** [comment]
Not directly, but uses predicted logits as features.

**Comment by ayush1801** [comment]
So, can we manually feed in the class labels for the applications like re-painting etc. Sorry, I didn't understand how you are feeding the class label into the Generator.

**Comment by rosinality** [comment]
Sorry I made mistake. Class labels from dataset is fed to generator using adaptive batch norm. And logits from vgg additionally used as features like feature maps from conv layers.

**Comment by ayush1801** [comment]
Thanks @rosinality ! So in all- inputs are noise vector, class labels fed through adaptive batch norm and the features (masked or unmasked) from the pre-trained classifier.

**Comment by rosinality** [comment]
Yes!

**Comment by ayush1801** [comment]
Thanks for your comments!

**Comment by ayush1801** [comment]
If the labels are provided by the Adaptive Batch norm, Doesn't that mean if we put self.norm=False, class labels will not be fed?

**Comment by ayush1801** [comment]
@rosinality  Could you please tell what is final resolution size of the image generated from the Generator, I am not able to figure out the same. Is that 224*224 only?

**Comment by rosinality** [comment]
About batch norm, yes, if you set norm=False, then labels will not be used in generator in current imlementations.

I have hard coded images sizes to be 224, and it seems like that authors also used 224px. You can increase image sizes, but I don't know how will vgg behave at higher resolutions.

**Comment by ayush1801** [comment]
Ok so generation size is also 224x224.

**Comment by rosinality** [comment]
Yes, it should be same with input image sizes due to reconstruction loss.

**Comment by ayush1801** [comment]
Thanks!

**Comment by ayush1801** [comment]
@rosinality I was curious about the way you passed the class_ids in the form of 128 dimensional embeddings! I am thinking that since they are passed independently to the architectures in different times, how the model differentiate between the classes?

**Comment by rosinality** [comment]
Model will learn to use and differentiate class informations as there are supervisions from feature matching with images corresponds to class labels, and projection discriminator that uses class informations for adversarial training.

**Comment by ayush1801** [comment]
@rosinality  Can you suggest what deletions/additions should be done if I want to run your code on google colab. It is showing CUDA out of memory layer. Also, it would be helpful if you could enlighten about the significance of accumulate() function for adding two model parameters. Thanks!

**Comment by rosinality** [comment]
If you want to training I think reducing batch sizes will be most simple way to reduce memory consumptions, but it will hurt the performance of the model. accumulate function will make ema of the model, and in many cases using ema is crucial for gan performances.

**Comment by ayush1801** [comment]
Thanks for the inputs!

**Comment by ayush1801** [comment]
@rosinality I found that reducing the number of iterations (after removing  EMA) model  started training. Do you how it was able to allocate memory after reducing number of iterations?

**Comment by rosinality** [comment]
Hmm the number of iterations should be irrelevant to the memory consumptions. Also I can't understand well the meaning of "allocate memory after reducing number of iterations".

**Comment by ayush1801** [comment]
Actually previous error was CUDA not able to allocate memory. After, reducing the iterations, it didn't give any error! Also, I am not able to understand when I am putting norm=True, it is showing error related to discriminator inputs - ' new(): argument 'size' must be tuple of ints, but found element of type NoneType at pos 2'. If you have any insights, it would be helpful!

**Comment by rosinality** [comment]
Where did you modified to be norm=True? Also, could give me a full error logs?

**Comment by ayush1801** [comment]
Sure I'll give you in a sec!
![image](https://user-images.githubusercontent.com/51224807/90669549-74614380-e26f-11ea-910b-080309170c7a.png)


![image](https://user-images.githubusercontent.com/51224807/90669422-43810e80-e26f-11ea-8c8a-395c22cd5ca5.png)

**Comment by rosinality** [comment]
As batch normalization used in resblock is adaptive one, it requires additional inputs like class conditions. So you can't use it in discriminator.

**Comment by ayush1801** [comment]
In my use case, images have to be formed by keeping the class_id (as per Conditional GANs). You told before if I didn't pass norm==True, generator would not learn to produce the images of the particular class_id. Please elaborate on this. Also, what do you mean by 'class conditions' - class_ids? It would be of great help!

**Comment by rosinality** [comment]
Yes, in this case class conditions injected using class ids. And by default generator uses class ids as norm=True in generator. (https://github.com/rosinality/semantic-pyramid-pytorch/blob/master/model.py#L316)

**Comment by ayush1801** [comment]
Ahh! Thanks so I don't need to put norm==True in the dis.

**Comment by rosinality** [comment]
Discriminator uses class informations using projection discriminator, so you don't need it.

**Comment by ayush1801** [comment]
@rosinality When I am training the model. The logs are printed twice! Is there a reason for such output? I am attaching the ss. Also, after 69th epoch, logs are printed in a different manner. Any info would be helpful!
![image](https://user-images.githubusercontent.com/51224807/90918701-c9818e80-e402-11ea-8e76-ff9c99b06a41.png)
![image](https://user-images.githubusercontent.com/51224807/90918764-de5e2200-e402-11ea-8d70-6632fb04cb9d.png)

**Comment by rosinality** [comment]
Maybe your terminal doesn't support the functionalities for tqdm. You can just replace tqdm with regular print function.

Also, second problem, I don't know why it happens. As tqdm and train loop iterates over fixed iterations, (set py --iter arguments) it can't iterate more than predefined iterations. (in your cases, 69 iteration) Something is wrong but hard to know why.

**Comment by ayush1801** [comment]
Thanks for the info! Also, it would be great to know when you stopped training the model as we know there are four losses! Did you observe all of them together/plotted graphs at regular intervals or looked only at minimizing G and D losses?

**Comment by rosinality** [comment]
I stopped training when I found model was collapsed. Maybe because loss calculations were wrong that times, or batch size was small. (64 vs 1024) And G and D losses itself will not be very informant.

**Comment by ayush1801** [comment]
Ok! Are the losses rectified now? Or are there still some issues? Also, I didn't understand why your model collapsed since the output shown in your readme file is as promising as given in the paper. Also, I wanted to ask that should we look of minimizing both the G and D losses or other things are required to analyse the training process?

**Comment by rosinality** [comment]
Loss calculation was fixed, and the sample of the readme file is from the model that before the collapse. And not both G and D losses will be minimized or decreased. (I don't remember which of it was decreased.) And it will be not very correleated with sample qualities. Maybe reconstruction loss could be meaningful, and it could be better to use FID as shown in the paper.

**Comment by ayush1801** [comment]
One more doubt! In the loss description you have d (I guess for discriminator), g (I guess for Generator), rec for reconstruction and div for diversity loss. You have also, calculated a_loss (g_ls_loss) and printed it in g loss. Also for both d and div you are printing d_loss (formula used for diversity loss). Also, for discriminator training you are using d_ls_loss and for Generator training you used g_loss (a_loss + r_loss + d_loss). Please see to this if any changes are needed!

**Comment by rosinality** [comment]
Ah, there was a bug. Fixed it at be3cb40. Anyway, you don't need any changes.

**Comment by ayush1801** [comment]
Thanks! Also, for generator you are training using g_loss (a_loss+rec+rec) not a_loss right? So, shouldn't logs show g_loss instead of a_loss?

**Comment by rosinality** [comment]
I think showing individual losses is easy to check training progresses. Anyway, you can add sum of the losses in the progress bar.

**Comment by ayush1801** [comment]
I think I was not able to explain my doubt - you are training the Generator using the loss which is the sum of g_ls_loss, reconstruction and diversity losses right? Not using the individual g_ls_loss ( denoted by a_loss)?

**Comment by rosinality** [comment]
Yes, generator is trained with 3 losses.

**Comment by ayush1801** [comment]
Fine Thanks for your inputs!

**Comment by ayush1801** [comment]
@rosinality Can you throw some light about the masking process you are using during training and during producing the final results? Are these as per the paper or you have put in your own thoughts to it? Thanks

**Comment by rosinality** [comment]
I have tried to match masking process with the paper. So you can refer to it.

**Comment by ayush1801** [comment]
Thanks!

**Comment by ayush1801** [comment]
Also, during saving the models you are saving EMA generator, g_module and d_module. Shouldn't it be gen and dis in place of g_module and d_module since gen and dis are the instances getting trained?

**Comment by rosinality** [comment]
As generator and discriminator object is wrapped with DistributedDataParallel, so saving module under the DistributedDataParallel is more convenient when load checkpoints.

**Comment by ayush1801** [comment]
Sorry, I didn't understand your answer. For your information, I am not using multi gpus and I have removed EMA also. I noticed that g_module and d_module are initialised before the loop of iterations so they might not reflect the updated weights during the training. Take a look at the snippet atttached.


def train(args, dataset, gen, dis, g_ema, device):
      if args.distributed:
          g_module = gen.module
          d_module = dis.module
      else:
          g_module = gen
          d_module = dis

But the updating is done on gen and dis instances and saving is done using g_module and d_module. Let me know if I am missing some logic!

**Comment by rosinality** [comment]
g_module and d_module is just reference of the generator or discriminator objects, and it is not a separated copy of it.

**Comment by ayush1801** [comment]
Thanks! Got it

**Comment by ayush1801** [comment]
One Problem! I am using 16 batch size. Till 24000 iterations, images generated were fine but after that iterations some images started showing random black patches (one image attached). Could you enlighten how this might be happening as it started occurring abruptly! Or shall I train till more iterations (currently trained till 64000 iterations) and have some patience? Tweak learning rates or other hypers? 
![iterno064000batchno6classid1](https://user-images.githubusercontent.com/51224807/92166792-d8caf800-ee56-11ea-91b8-a4fe8f6142d4.png)

**Comment by rosinality** [comment]
Training is collapsed. Note that in the paper large (1024) batch size is used. I think training with batch size 16 will be instable.

**Comment by ayush1801** [comment]
@rosinality  I was able to regain stable training by decreasing the learning rate after 24000 iterations from 1e-4 to 1e-5 for both the G and D. I just wanted to confirm one thing - is it safe to put the same label to the generated image as of the input image (mapping the classid of the input image to the classid to the generated image)? Great to have your suggestions..

**Comment by rosinality** [comment]
If you mean class id assignment to the discriminator, then it is genuine projection discriminator formulation.

**Comment by ayush1801** [comment]
I mean to say that if we supply classid=7 (Cat image) to the generator, is it safe to label the generated image (fake image) as a cat image (class_id=7) itself. In all, is your architecture as par with conditional GAN working like that of the paper? Thanks in advance

**Comment by rosinality** [comment]
It's upto your decision. Generally I think it could generate images the similar to the images in the classes. But it could be not enough for some applications.

**Comment by ayush1801** [comment]
@rosinality What changes should be done (a rough idea) if I want to use only the grayscale images (channel = 1 and not channel = 3)? Do we need to do many changes? Let me know. Thanks in advance!

**Comment by rosinality** [comment]
Most easy way will be change your images to rgb. Or you can change output/input channel of generator/discriminator to 1 and convert image channel to 3 before using vgg.

**Comment by ayush1801** [comment]
So the best thing would be to use to convert('RGB') -copying the single channel thrice since we have a pre-trained VGG model trained on Imagenet data (3-channeled) right?

**Comment by rosinality** [comment]
Yes.

**Comment by ayush1801** [comment]
@rosinality Can you please explain the use of classid embeddings in the Generator, I mean the portion where you used block(out,embed). It would be of great help!
https://github.com/rosinality/semantic-pyramid-pytorch/blob/be3cb40dc8a2f8dd92295628db77265fbe6be45e/model.py#L376

**Comment by rosinality** [comment]
It is used to modulate feature maps to make it dependent or conditioned on class information/conditions.

**Comment by ayush1801** [comment]
Ok so this is for the conditioning of the model in the form of a Conditional GAN for generating images of the particular label
